src/router_core/connections.c

```diff
@@ -1702,16 +1708,24 @@ static void qdr_link_inbound_first_attach_CT(qdr_core_t *core, qdr_action_t *act
     qdr_connection_t      *conn = safe_deref_qdr_connection_t(action->args.connection.conn);
     qdr_link_t            *link = safe_deref_qdr_link_t(action->args.connection.link);
     qdr_delivery_t *initial_dlv = action->args.connection.initial_delivery;
+
+    qdr_terminus_t *source = action->args.connection.source;
+    qdr_terminus_t *target = action->args.connection.target;
+
     if (discard || !conn || !link) {
         if (initial_dlv)
             qdr_delivery_decref(core, initial_dlv,
                                 "qdr_link_inbound_first_attach_CT - discarding action");
+        if (discard) {
+            qdr_terminus_free(source);
+            qdr_terminus_free(target);
+            qdr_link_free_CT(link);
```

Comment:
Freeing the link doesn't seem safe to me: the callers to qdr_link_inbound_first_attach() get a pointer to the link returned to them, so I'm pretty sure freeing this link here will result in... issues.

Reply:
Agreed. Ideally I would like to drop this link into core->open_links so the shutdown process will clean it up. But it looks like the qdr_core_free() first cleans up all actions and then later cleans up  core->open_links. Sigh.